### PR TITLE
add vike-react-apollo

### PR DESCRIPTION
**Pull request by nitedani**

### CODE REVIEW

**Comment by brillout** [comment]
:rocket:

Looking forward to merge :ok_hand:

You labeled it as draft so I guess it isn't ready for review yet, but let me know if you want a review.

**Comment by nitedani** [comment]
https://github.com/apollographql/apollo-client-nextjs/issues/325

**Comment by nitedani** [comment]
Figured it out 😛

**Comment by nitedani** [comment]
The PR "works fine" right now, but it's not perfect. The last chunk is not injected to the stream, because the stream is already closed. It surprisingly doesn't cause any issues with my primitive examples.

There are 3 ways to solve it - hardest to easiest:
1. https://github.com/brillout/react-streaming/issues/40
2. add a hook to `vike-react`, allowing wrapping the stream returned by `react-streaming` with https://github.com/apollographql/apollo-client-nextjs/blob/e7a59cb26716a77bbfac659f435f89f5af8eff61/packages/client-react-streaming/src/stream-utils/createInjectionTransformStream.tsx#L18C17-L18C47
3. use my custom `StreamedHydration` component instead of the official `@apollo/client-react-streaming` package, which has [this issue](https://github.com/apollographql/apollo-client-nextjs/issues/325) - though I think that could cause some other issues in the future
4. leave it as-is, and fix it when we encounter an issue in the future

**Comment by brillout** [comment]
> 1. [Allow passing async function to injectToStream brillout/react-streaming#40](https://github.com/brillout/react-streaming/issues/40)

If this is a reliable solution, then I'd be :+1: for doing this.

> 2. add a hook to `vike-react`, allowing wrapping the stream returned by `react-streaming` with https://github.com/apollographql/apollo-client-nextjs/blob/e7a59cb26716a77bbfac659f435f89f5af8eff61/packages/client-react-streaming/src/stream-utils/createInjectionTransformStream.tsx#L18C17-L18C47

That could be nice as well :+1:. Some kind of generic hook enabling the user to manipulate the stream.

> 4. (bonus) leave it as-is, and fix it when we encounter an issue in the future🥹

Let's see if we need go for the bonus :grin: (or `3.`)

**Comment by brillout** [comment]
```
$ npm owner add nitedani vike-react-apollo
npm notice INFO: User nitedani invited to package vike-react-apollo successfully.
+ nitedani (vike-react-apollo)
```

File: packages/vike-react-apollo/README.md [hunk 1]
```diff
@@ -0,0 +1,130 @@
+<!-- WARNING: keep links absolute in this file so they work on NPM too -->
+
+[<img src="https://vike.dev/vike-readme.svg" align="right" height="90">](https://vike.dev)
+[![npm version](https://img.shields.io/npm/v/vike-react-apollo)](https://www.npmjs.com/package/vike-react-apollo)
+
+# `vike-react-apollo`
```
Comment (brillout) [inline packages/vike-react-apollo/README.md:6]: I did some polishing for the `vike-react-query` docs: https://github.com/vikejs/vike-react/tree/main/packages/vike-react-query#readme. Maybe we can also apply some of that polish here as well.

File: packages/vike-react-apollo/utils/assertUsage.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
+export { assert, assertUsage }
+
+function assert(condition: unknown): asserts condition {
+  if (condition) return
+  throw new Error(
+    "You stumbled upon a bug in vike-react-apollo's source code. Reach out on GitHub and we will fix the bug."
```
Comment (brillout) [inline packages/vike-react-apollo/utils/assertUsage.ts:6]:
```suggestion
    "You stumbled upon a bug in vike-react-apollo's source code, reach out on GitHub."
```

File: packages/vike-react-apollo/README.md [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+   import type { Config } from 'vike/types'
+   import vikeReact from 'vike-react/config'
+   import vikeReactApollo from 'vike-react-apollo/config'
+
+   export default {
+     ...
+     extends: [vikeReact, vikeReactApollo]
+   } satisfies Config
+   ```
+3. Create `+ApolloConfig.ts`:
```
Comment (brillout) [inline packages/vike-react-apollo/README.md:41]: This currently only defines the Apollo client. If `new ApolloClient()` doesn't accept any other argument then maybe we can rename the setting to `+ApolloClient`.

In order to keep things as zero-config as possbile, maybe be can also support this for those who don't need `pageContext`:

```ts
// pages/+ApolloConfig.ts

import { InMemoryCache } from '@apollo/client-react-streaming'

export default {
  uri: 'https://countries.trevorblades.com',
  cache: new InMemoryCache()
}
```

Do we actually foresee a use case for using `pageContext` here? If we don't foresee one, then I'd say let's remove it and add it only after a user complains.

File: packages/vike-react-apollo/README.md [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+      }
+    )
+
+    return (
+      <div>
+        Name: <b>{data.country.name}</b>
+      </div>
+    )
+  },
+  {
```
Comment (brillout) [inline packages/vike-react-apollo/README.md:120]: Maybe we can also document `useFallback(() => {}, 'Loading...')` (the second argument not being an object but directly the loading component).

File: packages/vike-react-apollo/renderer/+config.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
+import type { Config } from 'vike/types'
+import type _ from 'vike-react/config' // Needed for declaration merging of Config
+import type { ApolloClientOptions } from './types.js'
+
+export default {
+  name: 'vike-react-apollo',
+  require: {
+    'vike-react': '>=0.4.13'
```
Comment (brillout) [inline packages/vike-react-apollo/renderer/+config.ts:8]:
```suggestion
    'vike-react': '>=0.4.178'
```

Comment (brillout) [reply packages/vike-react-apollo/renderer/+config.ts:8]: Ah, no, I mixed it up ^^

File: packages/vike-react-apollo/renderer/Transport.tsx [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    useInsertHtml() {
+      const stream = useStream()
+      if (!stream) {
+        return () => {}
+      }
+      return (callback: () => React.ReactNode) => {
+        stream.injectToStream(
+          // https://github.com/apollographql/apollo-client-nextjs/issues/325
+          new Promise((resolve) =>
+            setTimeout(() => {
```
Comment (brillout) [inline packages/vike-react-apollo/renderer/Transport.tsx:18]: Do we still need `setTimeout()` here? Doesn't seem right :thinking:

File: packages/vike-react-apollo/renderer/Wrapper.tsx [hunk 7]
```diff
@@ -0,0 +1,16 @@
+import { ApolloClient } from '@apollo/client-react-streaming'
+import React, { type ReactNode } from 'react'
+import { usePageContext } from 'vike-react/usePageContext'
+import { assertUsage } from '../utils/assertUsage.js'
+import { WrappedApolloProvider } from './Transport.js'
+
+export default function Wrapper({ children }: { children: ReactNode }) {
+  const pageContext = usePageContext()
+  const { ApolloConfig } = pageContext.config
+  assertUsage(ApolloConfig, 'ApolloConfig is required in config')
```
Comment (brillout) [inline packages/vike-react-apollo/renderer/Wrapper.tsx:10]:
```suggestion
  assertUsage(ApolloConfig, 'Setting +ApolloConfig is required')
```

File: packages/vike-react-apollo/src/withFallback.tsx [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+type WithFallbackOptions<T> = {
+  Loading?: Loading<T>
+  Error?: Error<T>
+}
+
+export function withFallback<T extends object = Record<string, never>>(
+  Component: ComponentType<T>,
+  options?: WithFallbackOptions<T>
+): ComponentType<T>
+export function withFallback<T extends object = Record<string, never>>(
```
Comment (brillout) [inline packages/vike-react-apollo/src/withFallback.tsx:24]: I feel like the third overload isn't worth it? I feel like it adds more noise than adding benefits?

File: packages/vike-react-apollo/src/withFallback.tsx [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+type Error<T> = ComponentType<T & ErrorFallbackProps> | ReactNode
+
+type WithFallbackOptions<T> = {
+  Loading?: Loading<T>
+  Error?: Error<T>
+}
+
+export function withFallback<T extends object = Record<string, never>>(
+  Component: ComponentType<T>,
+  options?: WithFallbackOptions<T>
```
Comment (brillout) [inline packages/vike-react-apollo/src/withFallback.tsx:22]: Actually, instead of having an object, how about we support `withFalllback(Component, null, Error)`?

**Comment by brillout** [comment]
Review done.

File: packages/vike-react-apollo/README.md (hunk 3 above)
Comment (nitedani) [reply packages/vike-react-apollo/README.md:41]: It needs to be a function, because the config file is only loaded once, but we need a new cache for each request server-side. PageContext is useful for proxying headers/cookies for authentication. For example the user can add an `ApolloLink` that reads the headers/cookies from the `pageContext` and sets it on the graphql request server-side.

File: packages/vike-react-apollo/src/withFallback.tsx (hunk 8 above)
Comment (nitedani) [reply packages/vike-react-apollo/src/withFallback.tsx:24]: There is only 1 overload, I don't understand.

File: packages/vike-react-apollo/src/withFallback.tsx (hunk 9 above)
Comment (nitedani) [reply packages/vike-react-apollo/src/withFallback.tsx:22]: It's supported already. Should we remove support for the object and only keep `withFalllback(Component, null, Error)`?

File: packages/vike-react-apollo/README.md (hunk 3 above)
Comment (brillout) [reply packages/vike-react-apollo/README.md]: I see, makes sense.

How about this:

```js
import { InMemoryCache, ApolloClient } from '@apollo/client-react-streaming'

export default (pageContext) =>
  new ApolloClient({
    uri: 'https://countries.trevorblades.com',
    cache: new InMemoryCache()
  })
```

File: packages/vike-react-apollo/src/withFallback.tsx (hunk 8 above)
Comment (brillout) [reply packages/vike-react-apollo/src/withFallback.tsx:24]: Ah, sorry, you're right.

File: packages/vike-react-apollo/src/withFallback.tsx (hunk 9 above)
Comment (brillout) [reply packages/vike-react-apollo/src/withFallback.tsx:22]: > It's supported already. Should we remove support for the object and only keep `withFalllback(Component, null, Error)`?

I'd be :+1: for removing support for passing an object. (Option 1.)

Alternatively (option 2), we could support only passing an object for defining `Error`:

```js
// Not supported
withFalllback(Component, null, Error)

// Supported
withFalllback(Component, { Error })
withFalllback(Component, Loading)
withFalllback(Component, { Loading, Error })
```

But, so far, I prefer option 1.

I'm thinking that supporting both option 1 and option 2 adds more noise than adds benefit.

**Comment by nitedani** [comment]
ready

File: packages/vike-react/src/renderer/onRenderClient.tsx [hunk 10]
```diff
@@ -5,12 +5,15 @@ import ReactDOM from 'react-dom/client'
...
 import type { OnRenderClientSync } from 'vike/types'
 import { getPageElement } from './getPageElement.js'
+import './styles.css'
 
 let root: ReactDOM.Root
 const onRenderClient: OnRenderClientSync = (pageContext): ReturnType<OnRenderClientSync> => {
   // Use case:
   // - Store hydration https://github.com/vikejs/vike-react/issues/110
-  pageContext.config.onBeforeRenderClient?.(pageContext)
+  for (const onBeforeRenderClient of pageContext.config.onBeforeRenderClient || []) {
```
Comment (brillout) [inline packages/vike-react/src/renderer/onRenderClient.tsx:14]: :+1: Neat, although let's do it in another PR. Technically, it's a (small) breaking change (but I think it's fine we release it in a patch).

File: packages/vike-react/src/renderer/styles.css [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+  max-height: 100%;
+  background: #eee;
+  background: linear-gradient(110deg, #ececec 8%, #f5f5f5 18%, #ececec 33%);
+  border-radius: 5px;
+  background-size: 200% 100%;
+  animation: 1.3s vike-react-apollo-shine linear infinite;
+  aspect-ratio: 2.5/1;
+}
+
+@keyframes vike-react-apollo-shine {
```
Comment (brillout) [inline packages/vike-react/src/renderer/styles.css:13]: The name should be updated.

File: packages/vike-react/src/renderer/styles.css [hunk 12]
```diff
@@ -0,0 +1,17 @@
+.vike-react-apollo-loading-component {
```
Comment (brillout) [inline packages/vike-react/src/renderer/styles.css:1]: How about we define this as an inline style?

The neat thing would also be that it doesn't get loaded if the user overrides `+LoadingComponent`.

File: packages/vike-react-apollo/package.json [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+  "name": "vike-react-apollo",
+  "version": "0.0.1",
+  "type": "module",
+  "main": "dist/src/index.js",
+  "typings": "dist/src/index.js",
+  "exports": {
+    ".": "./dist/src/index.js",
+    "./config": "./dist/renderer/+config.js",
+    "./renderer/Wrapper": "./dist/renderer/Wrapper.js",
+    "./renderer/ApolloConfig": "./dist/renderer/ApolloConfig.js"
```
Comment (brillout) [inline packages/vike-react-apollo/package.json:11]: This should be removed?

File: packages/vike-react-apollo/package.json [hunk 14]
```diff
@@ -0,0 +1,64 @@
+{
+  "name": "vike-react-apollo",
+  "version": "0.0.1",
```
Comment (brillout) [inline packages/vike-react-apollo/package.json:3]: I think we can release it directly as `0.1.0`.

**Comment by brillout** [comment]
Reviewing done.
- I did some minor refactoring (feel free to disagree)
- Root README.md needs to be updated
- https://vike.dev/Loading
- https://vike.dev/LoadingComponent

**Comment by brillout** [comment]
Idea:
```js
// +Loading.js

export default {
  Layout: SomeLoadingComponent, // for both the Page and all Layouts
  Component: SomeOtherLoadingComponent
}

// Sets Loading.Component
export default SomeOtherLoadingComponent
```

WDYT?

**Comment by nitedani** [comment]
I like it because `+LoadingComponent` sounds weird, `+Loading` doesn't.
But I don't know.

>// Sets Loading.Component
export default SomeOtherLoadingComponent

🧐

**Comment by brillout** [comment]
The way I see it is that's it's similar to overload function arguments.

```js
setLoading(value) // What should this do?
setLoading({ component: value, layout: otherValue })
```

I'm inclined to make `setLoading(value)` be a shortcut for setting `setLoading({ component: value })`. Same for `+Loading.js`.

But maybe there is a better interface looming around the corner :thinking:

**Comment by brillout** [comment]
Actually, I'm not sure: what should be the default page transition animation? I think we should eventually add one instead of just showing a blank page. I'm asking because maybe we want to make `export default SomeLoadingComponent` apply to both components and layouts :thinking:

Or how about this: we only accept `export default { component: Component, layout: Component }` for now, and decide later what `export default Component` should do. So that we can proceed with this PR.

Unrelated: how about  `{ components: Component, layouts: Component }`? (Added `s` to the option names.)

**Comment by brillout** [comment]
> Actually, I'm not sure: what should be the default page transition animation? I think we should eventually add one instead of just showing a blank page.

How about the "main Suspense approach" described at https://github.com/vikejs/vike-react/issues/117? So far I like this most. WDYT?

**Comment by nitedani** [comment]
> `export default { component: Component, layout: Component }` for now

done

**Comment by brillout** [comment]
You're right, without `s` is better.

**Comment by brillout** [comment]
LGTM. Feel free to merge & release. We need to manually squash the PR in two commits: one for `vike-react` and one of `vike-react-apollo`.
